pydantic/type_adapter.py

```diff
@@ -558,6 +554,9 @@ def json_schema(
             The JSON schema for the model as a dictionary.
         """
         schema_generator_instance = schema_generator(by_alias=by_alias, ref_template=ref_template)
+        if isinstance(self.core_schema, _mock_val_ser.MockCoreSchema):
```

Comment:
If the `core_schema` would have been a cached prop then we wouldnt need sprinkle these checks in the usage side

Reply:
This is true, but we wanted to model the pattern used for `BaseModel` as closely as possible:

https://github.com/pydantic/pydantic/blob/ff4a325cb1a1d4e00584366d086c652463c1ea35/pydantic/json_schema.py#L2348-L2359

Reply:
Ok I see!

Reply:
We dont need similar mock checks for others like `self.serializer` eg in `dump_json`?

Reply:
Good question, I don't think so, given that the `MockCoreSchema` logic supports rebuilds.

That makes me wonder if we actually need to be this complicated above, though.

Reply:
Might be within the scope of another PR to simplify that across the board...

Reply:
Ah yes, it's bc we're not calling `getattr` on `core_schema` like we would for `serializer`. That `getattr` is what triggers the rebuild.